trulens_eval/trulens_eval/feedback/provider/cortex.py

```diff
@@ -0,0 +1,105 @@
+
+from typing import ClassVar, Dict, Optional, Sequence
+
+import os
+import json
+from trulens_eval.feedback.provider.base import LLMProvider
+from trulens_eval.feedback.provider.endpoint.cortex import CortexEndpoint
+from trulens_eval.utils.imports import REQUIREMENT_CORTEX, OptionalImports
+
+with OptionalImports(messages=REQUIREMENT_CORTEX):
+    from snowflake.snowpark import Session
+
+
+class Cortex(LLMProvider):
+    # require `pip install snowflake-snowpark-python` and a active Snowflake account with proper privileges
+    # https://docs.snowflake.com/en/user-guide/snowflake-cortex/llm-functions#availability
+
+    DEFAULT_MODEL_ENGINE: ClassVar[str] = "snowflake-arctic"
+
+    model_engine: str
+    """Snowflake's Cortex COMPLETE endpoint. Defaults to `snowflake-arctic`.
+       Reference: https://docs.snowflake.com/en/sql-reference/functions/complete-snowflake-cortex
+    """
+
+    endpoint: CortexEndpoint
+    snowflake_session: Session
+
+    def __init__(
+        self,
+        model_engine: Optional[str] = None,
+        snowflake_connection_params: Optional[Dict] = None,
+        *args,
+        **kwargs: dict
+    ):
+        self_kwargs = dict(kwargs)
+
+        self_kwargs['model_engine'] = self.DEFAULT_MODEL_ENGINE if model_engine is None else model_engine
+        self_kwargs["endpoint"] = CortexEndpoint(
+            *args, **kwargs
+        )
+
+        connection_params = {
+            "account": os.environ["SNOWFLAKE_ACCOUNT"],
+            "user": os.environ["SNOWFLAKE_USER"],
+            "password": os.environ["SNOWFLAKE_USER_PASSWORD"],
+        } if snowflake_connection_params is None else snowflake_connection_params
+
+        # Create a Snowflake session
+        self_kwargs['snowflake_session'] = Session.builder.configs(
+            connection_params).create()
+
+        super().__init__(**self_kwargs)
+
+    def _escape_string_for_sql(self, input_string: str) -> str:
+        escaped_string = input_string.replace('\\', '\\\\')
+        escaped_string = escaped_string.replace("'", "''")
+        return escaped_string
+
+    def _exec_snowsql_complete_command(self, model: str, temperature: float, messages: Optional[Sequence[Dict]] = None):
+        # Ensure messages are formatted as a JSON array string
+        if messages is None:
+            messages = []
```

Comment:
I don't think it's possible to get here due to L88-L99

Reply:
ah yes good catch - removed